app/scripts/metamask-controller.js

```diff
@@ -6226,6 +6481,53 @@ export default class MetamaskController extends EventEmitter {
       origin = new URL(sender.url).origin;
     }
 
+    // solana account changed notifications
+    let caip25Caveat;
+    try {
+      caip25Caveat = this.permissionController.getCaveat(
+        origin,
+        Caip25EndowmentPermissionName,
+        Caip25CaveatType,
+      );
+    } catch {
+      // noop
+    }
+    if (caip25Caveat) {
+      const solanaAccountsChangedNotifications =
+        caip25Caveat.value.sessionProperties[
+          KnownSessionProperties.SolanaAccountChangedNotifications
+        ];
+
+      const sessionScopes = getSessionScopes(caip25Caveat.value, {
+        getNonEvmSupportedMethods: this.getNonEvmSupportedMethods.bind(this),
+      });
+
+      const solanaScope =
+        sessionScopes[MultichainNetworks.SOLANA] ||
+        sessionScopes[MultichainNetworks.SOLANA_DEVNET] ||
+        sessionScopes[MultichainNetworks.SOLANA_TESTNET];
+
+      if (solanaAccountsChangedNotifications && solanaScope) {
+        const { accounts } = solanaScope;
+        const parsedPermittedSolanaAddresses = accounts.map((caipAccountId) => {
+          const { address } = parseCaipAccountId(caipAccountId);
+          return address;
+        });
+
+        const accountAddressToEmit = this.sortMultichainAccountsByLastSelected(
+          parsedPermittedSolanaAddresses,
+        )[0];
+
+        if (accountAddressToEmit) {
+          // delay emit so that the event is not emitted
+          // before the wallet standard script is ready to receive it
+          setTimeout(() => {
+            this._notifySolanaAccountChange(origin, [accountAddressToEmit]);
+          }, 500);
```

Comment:
did we verify if this timeout can be removed if we make this come after the `this.addConnection()` that is below?

Reply:
Ah yea lets re-verify whether this is needed

Reply:
working on this

Reply:
https://github.com/MetaMask/metamask-extension/pull/29887/commits/fd60e752efa384647af18db8e4f41f81ee6df8fe

Reply:
@darkwing can you elaborate? are you referring to setInterval?